src/common/perf_counters_key.h

```diff
@@ -63,7 +63,9 @@ class label_iterator {
   reference operator*() const { return state->label; }
   pointer operator->() const { return &state->label; }
 
-  auto operator<=>(const label_iterator& rhs) const = default;
+  // can't request a default spaceship operator (<=>)
+  // because of the optional member `state`
+  bool operator==(const label_iterator& rhs) const = default;
```

Comment:
i don't understand this change. the commit message says
> The requested default spaceship operator for the type is not possible due to an optional member which doesn't support three way comparison

but std::optional does provide three-way comparison https://en.cppreference.com/w/cpp/utility/optional/operator_cmp, and so does struct iterator_state

what error does this cause?

Reply:
hmm, this is strange. It says "since C++20" on the cppreference, but apparently, the clang from macOs SDK 14.2 doesn't implement `<=>` for `std::optional` even if the type itself is `three_way_comparable`

I can update the comment to mention clang, and also some version macros to see when clang gets to implementing the standard.

```
[2/625] Building CXX object src/CMakeFiles/common-objs.dir/osd/PGPeeringEvent.cc.o
In file included from /Users/leonidus/projects/ibm/ceph/ceph/src/osd/PGPeeringEvent.cc:5:
In file included from /Users/leonidus/projects/ibm/ceph/ceph/src/osd/PGPeeringEvent.h:8:
In file included from /Users/leonidus/projects/ibm/ceph/ceph/src/osd/osd_types.h:53:
In file included from /Users/leonidus/projects/ibm/ceph/ceph/src/osd/osd_perf_counters.h:8:
/Users/leonidus/projects/ibm/ceph/ceph/src/common/perf_counters_key.h:68:8: warning: explicitly defaulted three-way comparison operator is implicitly deleted [-Wdefaulted-function-deleted]
  bool operator<=>(const label_iterator& rhs) const = default;
       ^
/Users/leonidus/projects/ibm/ceph/ceph/src/common/perf_counters_key.h:79:33: note: defaulted 'operator<=>' is implicitly deleted because there is no viable three-way comparison function for member 'state'
  std::optional<iterator_state> state;
                                ^
...
```

Reply:
I updated now to the latest xcode, and still the same issue.
Anyway, @cbodley, thanks for this comment! I neglected to check whether the optional _should_ support the three-way comparison and assumed it was GCC's extension.

For now, I will change this place to an `#if __APPLE__ ...` section. Later I'll consider adding a custom operator like i did [here](https://github.com/ceph/ceph/pull/52629/commits/25e05a4539e19c0138970482ec1a278908816c3a#diff-0832b60e2c472c0a3a99b0914754409598e9f3d38dd3446dea5ead4cf97f2bc9)

Reply:
i set up a godbolt example in https://godbolt.org/z/Tf73K4oY8 and saw that same failure with clang16/libc++. it succeeds with clang17/libc++

Reply:
Yeah.. I got it working by adding the below to the `convenience.h` and including it from the perf counters. Snatched it from the gcc's cppstdlib. 

```cpp
// despite https://en.cppreference.com/w/cpp/utility/optional/operator_cmp
// claiming optional's support for three way comparison since C++20
// as of xcode 15 (clang-1500.1.0.2.5) this is still not the case
template <typename _Tp, std::three_way_comparable_with<_Tp> _Up>
  requires(!requires(const std::optional<_Tp>& __x, const std::optional<_Up>& __y) { __x.operator<=>(__y); })
constexpr std::compare_three_way_result_t<_Tp, _Up>
operator<=>(const std::optional<_Tp>& __x, const std::optional<_Up>& __y)
{
  return __x && __y ? *__x <=> *__y : bool(__x) <=> bool(__y);
}
```